packages/mui-base/src/Slider/Root/useSliderRoot.ts

```diff
@@ -0,0 +1,448 @@
+        const distance = Math.abs(currentValue - value);
+
+        if (acc === null || distance < acc.distance || distance === acc.distance) {
+          return {
+            distance,
+            index,
+          };
+        }
+
+        return acc;
+      },
+      null,
+    ) ?? {};
+  return closestIndex;
+}
+
+export function focusThumb({
+  sliderRef,
+  activeIndex,
+  setActive,
+}: {
+  sliderRef: React.RefObject<any>;
+  activeIndex: number;
+  setActive?: (num: number) => void;
+}) {
+  const doc = ownerDocument(sliderRef.current);
+  if (
+    !sliderRef.current?.contains(doc.activeElement) ||
+    Number(doc?.activeElement?.getAttribute('data-index')) !== activeIndex
+  ) {
+    sliderRef.current?.querySelector(`[type="range"][data-index="${activeIndex}"]`).focus();
+  }
+
+  if (setActive) {
+    setActive(activeIndex);
+  }
+}
+
+function setValueIndex({
+  values,
+  newValue,
+  index,
+}: {
+  values: number[];
+  newValue: number;
+  index: number;
+}) {
+  const output = values.slice();
+  output[index] = newValue;
+  return output.sort(asc);
+}
+
+export function validateMinimumDistance(
+  values: number | readonly number[],
+  step: number,
+  minStepsBetweenValues: number,
+) {
+  if (!Array.isArray(values)) {
+    return true;
+  }
+
+  const distances = values.reduce((acc: number[], val, index, vals) => {
+    if (index === vals.length - 1) {
+      return acc;
+    }
+
+    acc.push(Math.abs(val - vals[index + 1]));
+
+    return acc;
+  }, []);
+
+  return Math.min(...distances) >= step * minStepsBetweenValues;
+}
+
+export function trackFinger(
+  event: TouchEvent | PointerEvent | React.PointerEvent,
+  touchIdRef: React.RefObject<any>,
+) {
+  // The event is TouchEvent
+  if (touchIdRef.current !== undefined && (event as TouchEvent).changedTouches) {
+    const touchEvent = event as TouchEvent;
+    for (let i = 0; i < touchEvent.changedTouches.length; i += 1) {
+      const touch = touchEvent.changedTouches[i];
+      if (touch.identifier === touchIdRef.current) {
+        return {
+          x: touch.clientX,
+          y: touch.clientY,
+        };
+      }
+    }
+
+    return false;
+  }
+
+  // The event is PointerEvent
+  return {
+    x: (event as PointerEvent).clientX,
+    y: (event as PointerEvent).clientY,
+  };
+}
+
+export const Identity = (x: any) => x;
+/**
+ *
+ * Demos:
+ *
+ * - [Slider](https://mui.com/base-ui/react-slider/#hooks)
+ *
+ * API:
+ *
+ * - [useSliderRoot API](https://mui.com/base-ui/react-slider/hooks-api/#use-slider-root)
+ */
+function useSliderRoot(parameters: UseSliderParameters): UseSliderReturnValue {
+  const {
+    'aria-labelledby': ariaLabelledby,
+    defaultValue,
+    direction = 'ltr',
+    disabled = false,
+    largeStep = 10,
+    max = 100,
+    min = 0,
+    minStepsBetweenValues = 0,
+    name,
+    onValueChange,
+    onValueCommitted,
+    orientation = 'horizontal',
+    rootRef,
+    scale = Identity,
+    step = 1,
+    tabIndex,
```

Comment:
@michaldudak `tabIndex` in the return value here gets put into context and is used by the Thumb like this: https://github.com/mui/base-ui/blob/55a30f74e5b276adb0d9ecf6beca2b149e5cc886/packages/mui-base/src/Slider/Root/SliderRoot.test.tsx#L1587-L1590

I suppose it could be passed straight from the `SliderRoot` component into the context without going through the hook, what do you think?

Reply:
Yup, if it's not used by the hook, it shouldn't go there.